### PR TITLE
GH-40968: [C++][Gandiva] add  RE2::Options set_dot_nl(true) for Like function

### PR DESCRIPTION
**Pull request by xxlaykxx**
### Rationale for this change

Gandiva function "LIKE" does not always work correctly when the string contains \n.
String value:
`[function_name: "Space1.protect"\nargs: "passenger_count"\ncolumn_name: "passenger_count" ]`
Pattern '%Space1%' nor '%Space1.%' do not match.

### What changes are included in this PR?

added flag set_dot_nl(true) to LikeHolder

### Are these changes tested?

add unit tests.

### Are there any user-facing changes?
Yes

**This PR includes breaking changes to public APIs.**

* GitHub Issue: #40968

### CODE REVIEW

**Comment by xxlaykxx** [comment]
@kou could you plz review?

**Review by kou — COMMENTED** [review]
@niyue @js8544 What do you think about this?

**Comment by kou** [comment]
Could you rebase on main to fix CI failures?

**Comment by js8544** [comment]
> Gandiva function "LIKE" does not always work correctly when the string contains \n.
String value:
[function_name: "Space1.protect" args: "passenger_count" column_name: "passenger_count" ]
Pattern '%Space1%' match string, but '%Space1.%' - not.

Could you please clarify the example? There are no '\n' in the target string. Because the pattern '%Space1.%' does match the string you give.

**Comment by js8544** [comment]
I do believe it is kind of a "bug" in Gandiva, a minimal reproducing example is `'abc\nd' LIKE '%abc%'` returns false.  I checked with mainstream dbs like Postgres, Redshift and Snowflake. They all return true for this case. But I'm not sure whether this is worth a breaking change.

File: cpp/src/gandiva/regex_functions_holder_test.cc [hunk 1]
```diff
@@ -46,6 +46,7 @@ class TestLikeHolder : public ::testing::Test {
 };
 
 TEST_F(TestLikeHolder, TestMatchAny) {
+  regex_op.set_dot_nl(true);
```
Comment (js8544) [inline cpp/src/gandiva/regex_functions_holder_test.cc:49]: Can this set_dot_nl be moved to the constructor, i.e. enable it for all test cases?

File: cpp/src/gandiva/regex_functions_holder_test.cc [hunk 2]
```diff
@@ -70,13 +72,23 @@ TEST_F(TestLikeHolder, TestMatchOne) {
...
   EXPECT_OK_AND_ASSIGN(auto const like_holder, LikeHolder::Make(".*ab_", regex_op));
 
   auto& like = *like_holder;
   EXPECT_TRUE(like(".*abc"));  // . and * aren't special in sql regex
   EXPECT_FALSE(like("xxabc"));
 }
 
+TEST_F(TestLikeHolder, TestPcreSpecialWithNewLine) {
+  regex_op.set_dot_nl(true);
+  EXPECT_OK_AND_ASSIGN(auto const like_holder, LikeHolder::Make("%Space1.%", regex_op));
```
Comment (js8544) [inline cpp/src/gandiva/regex_functions_holder_test.cc:85]: Can you also add a simpler test case, e.g. 'abc\nd' LIKE '%abc%' is enough to demonstate this change.

**Comment by kou** [comment]
> I do believe it is kind of a "bug" in Gandiva, a minimal reproducing example is `'abc\nd' LIKE '%abc%'` returns false. I checked with mainstream dbs like Postgres, Redshift and Snowflake. They all return true for this case. But I'm not sure whether this is worth a breaking change.

Oh. OK. Let's "fix" this without backward compatibility.

File: cpp/src/gandiva/regex_functions_holder.cc [hunk 3]
```diff
@@ -118,24 +119,13 @@ Result<std::shared_ptr<LikeHolder>> LikeHolder::Make(const FunctionNode& node) {
...
-  std::string pcre_pattern;
-  ARROW_RETURN_NOT_OK(RegexUtil::SqlLikePatternToPcre(sql_pattern, pcre_pattern));
-
-  auto lholder = std::shared_ptr<LikeHolder>(new LikeHolder(pcre_pattern));
-  ARROW_RETURN_IF(!lholder->regex_.ok(),
-                  Status::Invalid("Building RE2 pattern '", pcre_pattern,
-                                  "' failed with: ", lholder->regex_.error()));
-
-  return lholder;
-}
```
[the remark this answers is not in the document]
Comment (kou) [reply cpp/src/gandiva/regex_functions_holder.cc:135]: It seems that this is an exported API. If we remove this, we break backward compatibility. Is it expected?

[the remark this answers is not in the document]
Comment (xxlaykxx) [reply cpp/src/gandiva/regex_functions_holder.cc]: Good point. Restored

**Review by js8544 — APPROVED** [review]

**Review by kou — APPROVED** [review]
+1

> Gandiva function "LIKE" does not always work correctly when the string contains \n.
> String value:
> `[function_name: "Space1.protect"\nargs: "passenger_count"\ncolumn_name: "passenger_count" ]`
> Pattern '%Space1%' match string, but '%Space1.%' - not.

Is this description correct? I think that `%Space1%` doesn't match the string because the string includes `\n`.

**Comment by xxlaykxx** [comment]
@kou yes, this is how we found a problem. RE2 documentation said that  '.' any character, possibly including newline. If in pattern we don't use any special symbols - it work. But with special symbols and with \n in string - not. But flag handle this case.

**Comment by js8544** [comment]
> @kou yes, this is how we found a problem. RE2 documentation said that '.' any character, possibly including newline. If in pattern we don't use any special symbols - it work. But with special symbols and with \n in string - not. But flag handle this case.

kou means that `'[function_name: "Space1.protect"\nargs: "passenger_count"\ncolumn_name: "passenger_count" ]' LIKE  '%Space1%'` also returns false. So the statement that 

> Pattern '%Space1%' match string,

is not accurate.

**Review by niyue — APPROVED** [review]
+1
Besides the `LIKE` keyword, I tested postgres 16's `regexp_like` function, which matches the new line character by default as well

```
SELECT regexp_like(E'hello\nfoo', 'hello.*'); ==> true
```

**Comment by kou** [comment]
We use the description for commit message. So I want to use correct description.
Could you answer my question again? I meant what js8544 described https://github.com/apache/arrow/pull/40970#issuecomment-2046692241 . (Thanks, js8544!)

**Comment by xxlaykxx** [comment]
@kou it's already fixed

**Comment by kou** [comment]
Thanks.
I've merged.

**Pull request merged**